### ISSUE TITLE
data not from oracle in date 2020-03-21 (Daylight Saving Time)

### ISSUE THREAD

**Issue by mbtolou** [ClickHouse/ClickHouse#13845]
when i try to import data from oracle in 2020-03-21 , 30 % of records import and others ignored.

my clickhouse DateTime timezone is Asia-Tehran

**Comment by mbtolou** [comment]
Data of Oracle 

![image](https://user-images.githubusercontent.com/6413705/90461685-33a4e580-e11c-11ea-8070-8f1b553bfef4.png)

Data Imported in clickhouse

![image](https://user-images.githubusercontent.com/6413705/90461820-8da5ab00-e11c-11ea-8e31-f7b47c94cb05.png)


my table engine is **ReplacingMergeTree**
by SERVER_ID,DATA_ID,TV

in second image 3 record by same value in this columns(SERVER_ID,DATA_ID,TV).
and others record is missed.

**Comment by den-crane** [comment]
1. How did you get data from Oracle? 
2. What query did you use to insert data to CH?  
3. What is the table DDL?
4. Did you try  to use MergeTree instead of ReplacingMergeTree?
5. Can you provide data sample so I can reproduce your issue?
6. What CH version do  you use?

**Comment by mbtolou** [comment]
answer :

1.By C# driver
2.bulk insert
3.
```
CREATE TABLE MDM.BIZ_PUB_DATA_USE_T_S
(
    `SERVER_ID` UInt8,
    `DATA_ID` String,
    `TV` DateTime('Asia/Tehran'),
    `DELTA_A` Nullable(Decimal(12, 5)),
    `DELTA_R` Nullable(Decimal(12, 5)),
    `CT` Nullable(Decimal(12, 5)),
    `PT` Nullable(Decimal(12, 5)),
    `RECEIVE_TIME` DateTime('Asia/Tehran'),
    `USE_A` Nullable(Decimal(12, 5)),
    `USE_R` Nullable(Decimal(12, 5)),
    `TRG` Int8)
ENGINE = ReplacingMergeTree(TV)
PARTITION BY toYYYYMM(TV)
ORDER BY (TV, SERVER_ID, DATA_ID)
SETTINGS index_granularity = 8192;
```
4.yes
5.i write sample ,but work correctly. 
6.ClickHouse server version 20.6.3.28 (official build).

**Comment by mbtolou** [comment]
i found another problem : 

this query : 

```
SELECT  addHours(toDate('2020-03-21'),0) H1,addHours(toDate('2020-03-21'),1) H2
```

return

![image](https://user-images.githubusercontent.com/6413705/90468483-d7e25880-e12b-11ea-968e-f4eea06f3845.png)

**Comment by mbtolou** [comment]
I am using this driver : 
this lines fetch data  from IDataReader and insert into ClickHouse in raw binary format .

https://github.com/DarkWanderer/ClickHouse.Client/blob/c17fb5ca42bcb9d38e270d238fae50524dc1b6d4/ClickHouse.Client/Copy/ClickHouseBulkCopy.cs#L134

**Comment by mbtolou** [comment]
> i found another problem :
> 
> this query :
> 
> ```
> SELECT  addHours(toDate('2020-03-21'),0) H1,addHours(toDate('2020-03-21'),1) H2
> ```
> 
> return same result
> 
> ![image](https://user-images.githubusercontent.com/6413705/90468483-d7e25880-e12b-11ea-968e-f4eea06f3845.png)


SELECT toDateTime('2020-03-21 00:00:00'),toDateTime('2020-03-21 01:00:00'),toDateTime('2020-03-21 23:00:00')

![image](https://user-images.githubusercontent.com/6413705/90474147-e0da2680-e139-11ea-9997-9b2bbe24a898.png)

**Comment by den-crane** [comment]
> `SELECT toDateTime('2020-03-21 00:00:00'),toDateTime('2020-03-21 01:00:00'),toDateTime('2020-03-21 23:00:00')`
> 
> ![image](https://user-images.githubusercontent.com/6413705/90474147-e0da2680-e139-11ea-9997-9b2bbe24a898.png)

Do you mean it's a wrong day for the time shift ?

**Comment by mbtolou** [comment]
yes.
There is also another error ,that related to this error :
for example : 

```
CREATE TABLE test.USE_T
(
    `DATA_ID` String,
    `DATET` DateTime('Asia/Tehran'),
    `DELTA_A` Nullable(Decimal(12, 5))
)
ENGINE = ReplacingMergeTree(DATET)
PARTITION BY toYYYYMM(DATET)
ORDER BY (DATET,  DATA_ID)
SETTINGS index_granularity = 8192;


insert into test.USE_T
SELECT toString( toInt32(number)),addHours(toDate('2020-03-21','Asia/Tehran'),number%24),0
from numbers(128)

SELECT *
from test.USE_T ut 
where DATET =toDateTime('2020-03-21 00:00:00')
```

**Comment by den-crane** [comment]
>two last query return same date in result,in diffrence records.

Because they are two different hours in UTC rendered into the same String in Asia/Tehran because of timeshift.

**Comment by mbtolou** [comment]
> > `SELECT toDateTime('2020-03-21 00:00:00'),toDateTime('2020-03-21 01:00:00'),toDateTime('2020-03-21 23:00:00')`
> > ![image](https://user-images.githubusercontent.com/6413705/90474147-e0da2680-e139-11ea-9997-9b2bbe24a898.png)
> 
> Do you mean it's a wrong day for the time shift ?

This is normal alsoo ?

**Comment by alexey-milovidov** [comment]
@mbtolou I have validated the latest query
```
SELECT toDateTime('2020-03-21 00:00:00'),toDateTime('2020-03-21 01:00:00'),toDateTime('2020-03-21 23:00:00')
```
and I see that it is fixed in master.

It has been fixed in #9404
The change will be shipped in version 21.4 in April.
